cfep-10, feedstock support statuses

CJ-Wright opened the pull request:
@conda-forge/core I think this is ready for/needs discussion

closes https://github.com/conda-forge/conda-forge.github.io/issues/539

CJ-Wright committed:
cfep-10, feedstock support statues

scopatz commented:
I think there should be language added about the responsiveness to issues & PRs from real humans too, not just the bot

CJ-Wright commented on cfep-10.md:
```diff
@@ -0,0 +1,59 @@
+This proposal creates four new statuses
+
+### In Perfect Standing
+The feedstock is actively maintained, new versions of the packages are being created as needed, autotick bot PRs are addressed 
```
```suggestion
The feedstock is actively maintained, new versions of the packages are being created as needed, autotick bot and manual PRs and issues are addressed 
```

pkgw replied:
Could just write "... created as needed, PRs and issues are addressed ..."?

CJ-Wright commented on cfep-10.md:
```diff
@@ -0,0 +1,59 @@
+
+### In Perfect Standing
+The feedstock is actively maintained, new versions of the packages are being created as needed, autotick bot PRs are addressed 
+(merged, fixed, or closed) promptly. Less than 3 bot PRs are open at once.
```
```suggestion
(merged, fixed, or closed) promptly. Less than 3 PRs are open at once.
```

minrk replied:
Using a hard line like this encourages maintainers close user-contributed work-in-progress PRs. Is that a goal? Bot PRs represent tasks that should be done in a timely manner, user PRs may or may not be. I've seen a number of repos with stalled "build on Windows" or "split outputs" PRs outstanding for a long time, where their incompleteness is not an indication of the repo being unmaintained.

While I think a simple numerical metric for open bot PRs is valid, it's harder to have a good metric for user PRs (just counting open doesn't seem good to me, but e.g. open and passing tests without comment since last commit for a length of time...)

minrk replied:
I'd leave this statement as "Less than 3 bot PRs are open at once" and let community PRs be more vaguely addressed by the previous statement, relying on human judgment instead of bots to enforce "addressed promptly".

CJ-Wright commented on cfep-10.md:
```diff
@@ -0,0 +1,59 @@
+More maintainers are always welcome.
+
+### In Good Standing
+Maintainers are responsive but probably overworked. Multiple bot PRs have begun to pile up.
```
```suggestion
Maintainers are responsive but probably overworked. Multiple PRs and issues have begun to pile up.
```

CJ-Wright commented on cfep-10.md:
```diff
@@ -0,0 +1,59 @@
+This feedstock has no or unresponsive maintainers and either needs new maintainers or to be archived.
+Advertise that this needs maintainers and for the current ones re-engage with the feedstock or remove
+themselves, otherwise the feedstock will be archived. Set a deadline of X days/weeks/months for new
+maintainers to come on board, if nothing happens the feedstock is archived.
```
```suggestion
maintainers to come on board or existing maintainers to re-engage, if nothing happens the feedstock is archived.
```

minrk commented:
This is great! I see only two things missing:

1. how is standing communicated (maybe not in this PR, but perhaps worth discussing)? Is it a badge on the README? A listing on conda-forge.org?
2. Process for transitioning from one state to another. Time limits, state thresholds, warning & notification mechanisms, etc.

I think 1. warrants discussion, but perhaps doesn't need to be specified as part of this CFEP. 2. seems like it should be at least partly here, though, since there are a lot of questions about it, e.g. how automated can we make it? I don't see a great way to automatically threshold user interactions, but thresholding bot PRs is easy enough, though it might not be sufficient.

Questions if classification is automated:

1. what inputs are considered?
   1. is "activity" relevant? e.g. some CB3 pull requests were outstanding for a very long time, but were not inactive. They were just hard!
   2. are user-interactions (Issues, PRs) considered? I think this is really hard, but could be something like: number of issues where last activity is not from a maintainer and last comment is at least X time ago (a "known issue" label can put a hold on this to avoid forcing "still an Issue" comments). For PRs, ignore those with "WIP" in title
2. what are the thresholds?
   1. are they instantaneous (i.e. 3 bot PRs open now), or time-limited (3 bot PRs open for 3 weeks)
   2. what's the "unresponsive" threshold?
   3. what's the time limit for unresponsive -> archived?
   4. What sort of activity brings a repo back from unresponsive to good standing? Merging one PR? Making one comment on an Issue?

And if classification can have human input:

1. Is there a mechanism to *propose* transitions, but transitions are actually handled manually (by whom?). e.g. a user commenting "this package is broken and I'm not getting help, `ping @conda-forge/unresponsive-bot`"

CJ-Wright committed:
Update cfep-10.md

CJ-Wright committed:
Update cfep-10.md

CJ-Wright commented:
> how is standing communicated (maybe not in this PR, but perhaps worth discussing)? Is it a badge on the README? A listing on conda-forge.org?

I think this is an important implementation decision. I think we may have more than one place where this information is exposed. I definitely want it somewhere in the feedstock graph, since we could then have bots interact with it. It is also important to expose it to would be maintainers, most likely in a central place and to existing maintainers somewhere on the feedstock. I guess the would be maintainers could get something derived from the graph itself.

> Process for transitioning from one state to another. Time limits, state thresholds, warning & notification mechanisms, etc.

Yep, numbers need to be assigned to things (I just filled in things that seemed not unreasonable, but I have no claim that they are right or good).

 > is "activity" relevant? e.g. some CB3 pull requests were outstanding for a very long time, but were not inactive. They were just hard!

The bot currently tracks the PRs it issues with a high level of fidelity, I think we can go from the most recent update (push, comment, etc.).

> are they instantaneous (i.e. 3 bot PRs open now), or time-limited (3 bot PRs open for 3 weeks)

As a piece of context, the bot stop issuing version bumps after 3 have gone unmerged/closed.

> What sort of activity brings a repo back from unresponsive to good standing? Merging one PR?

I think that closing/merging some of the bot PRs would be good. Since we can have three at once, it is unlikely that all need to be merged, so closing would be a low cost way to get back into good standing.

CJ-Wright commented:
This also should include some discussion of feedstocks that have no maintainers.

CJ-Wright commented:
We might also consider what to do when a large number of version bump PRs are rejected without cause?

CJ-Wright commented:
@conda-forge/core
This PR falls under the CFEP policy, please vote and/or comment on this PR.
This PR needs 60% (12) of core to vote yea to pass.
To vote please leave Approve (yea) or Request Changes (nay) reviews.
If you would like changes to the current language please leave a comment or push to this branch.
This vote will end on Jan 22nd (2 weeks).

pkgw commented on cfep-10.md:
```diff
@@ -0,0 +1,59 @@
+<tr><td> Created </td><td> Sept 29, 2019</td></tr>
+<tr><td> Discussion </td><td> link to the PR where the CFEP is being discussed, NA is circulated initially </td></tr>
+<tr><td> Implementation </td><td> link to the PR for the implementation, NA if not availble </td></tr>
+</table>
```
Typos: `NA is circulated initially`, `availble`.

pkgw commented on cfep-10.md:
```diff
@@ -0,0 +1,59 @@
+This document proposes the creation of four new statuses for feedstocks that denote the	
+standing of the feedstock and transistions between the standings.
+These statuses will help us triage feedstocks and identify which feedstocks need
+help, maintainers, or archiving.
```
Should explicitly state that this CFEP does not "implement" these statuses in terms of saying how they should be surfaced, tracked, etc.

pkgw commented on cfep-10.md:
```diff
@@ -0,0 +1,59 @@
+help.
+Additionally, it is difficult for core to know which feedstocks need additional help or archiving.
+Finally, users of CF packages may find it difficult to understand which packages are dorment.
+This proposal aims to expose this information more clearly.
```
Typos: change to `would-be`, `dormant`.

pkgw replied:
The last sentence of this paragraph does not seem strictly correct. This proposal has very little to say about *exposing* status information at the moment.

pkgw commented on cfep-10.md:
```diff
@@ -0,0 +1,59 @@
+
+This proposal creates four new statuses
+
+### In Perfect Standing
```
Nothing in this world is perfect. "Excellent standing", perhaps?

pkgw commented on cfep-10.md:
```diff
@@ -0,0 +1,59 @@
+
+### In Perfect Standing
+The feedstock is actively maintained, new versions of the packages are being created as needed, autotick bot PRs are addressed 
+(merged, fixed, or closed) promptly. Less than 3 bot PRs are open at once.
```
Grammar: "fewer than three PRs are open".

pkgw commented on cfep-10.md:
```diff
@@ -0,0 +1,59 @@
+
+### In Good Standing
+Maintainers are responsive but probably overworked. Multiple PRs and issues have begun to pile up.
+We should advertise these as needing new maintainers.
```
For clarity, I think what is meant here is "needing *additional* maintainers", correct? "needing new maintainers" might be taken to suggest that the current ones should be kicked off, which I presume is not the intention.

pkgw commented on cfep-10.md:
```diff
@@ -0,0 +1,59 @@
+This feedstock has no or unresponsive maintainers and either needs new maintainers or to be archived.
+Advertise that this needs maintainers and for the current ones re-engage with the feedstock or remove
+themselves, otherwise the feedstock will be archived. Set a deadline of X days/weeks/months for new
+maintainers to come on board or existing maintainers to re-engage, if nothing happens the feedstock is archived.
```
Note that the sentence about setting a deadline is a piece of implementation policy, not really part of the definition of the "unsupported" status.

pkgw reviewed (changes requested):
I think this proposal needs refinement. Echoing some of the comments that have already been made, I think that proposals on this general topic should clearly distinguish:

- *Definitions* of feedstock support statuses, which might be qualitative or quantitative
- *Policies* such as triggers for transitioning from one status to another, when statuses are qualitative
- *Processes* for recording status, managing transitions, and surfacing status information

A CFEP that aims to cover all of these topics should get into the specifics a lot more than this current incarnation does. It's fine to keep things at the current high level, but in that case I think that the language about transitions, policies, and exposing the information should be removed — this CFEP should be purely definitional, and be followed with a future proposal laying out policies and processes.

CJ-Wright commented:
@pkgw would you please push to this branch?

pkgw commented:
@CJ-Wright Only the small-bore items that are straightforward to fix, or touching on all of my comments?

CJ-Wright commented:
I would do all of them you think appropriate.

More broadly I think this CFEP should cover the definitions and the policies. The implementation of these hopefully doesn't need to go through the CFEP process since those are technical details and are subject to change as new approaches come about. My hope is that once this is merged there will be enough details about desired outcomes that the technical implementation (eg. is this info on the status page, is there a badge on the feedstock, is there messaging when the package is downloaded, etc.) can be discussed in a faster medium.

pkgw committed:
cfep-10.md: some small-bore improvements:

- Add Discussion link in header
- NA for Implementation link in header
- Spelling fixes
- "Perfect Standing" to "Excellent Standing"
- For "excellent standing" feedstocks, specify that PRs and issues
  are addressed promptly, regardless of whether they're from bots or not.
- "new" to "additional" when describing adding maintainers

pkgw committed:
cfep-10.md: start trying to separate policies from definitions

pkgw commented:
OK, I pushed some changes that fixed the low-level stuff and tried to give an example of splitting out a set of policies (maintainer interactions) from the strict definitions of the statuses.

beckermr commented on cfep-10.md:
```diff
@@ -0,0 +1,70 @@
+
+<table>
+<tr><td> Title </td><td> Feedstock Statuses </td>
+<tr><td> Status </td><td> Proposed </td></tr>
```
```suggestion
<tr><td> Status </td><td> Deferred </td></tr>
```

beckermr committed:
Update cfep-10.md

beckermr commented:
merging as deferred
